feat(compiler-core): allow directive modifiers to be dynamic (fix #8281)

Fuzzyma opened the pull request:
fix #8281

This PR adds dynamic modifiers for directives 

```vue
<div v-foo.[myModifiers]="5" />

const myModifiers = {
  mod1: true,
  mod2: false
}
```

It also adds error states for empty modifiers (`v-foo.="5"`) and handles empty dynamic modifiers or invalid values e.g. `v-foo.[]="5"`, `v-foo.[123]="5"`.

I skipped modifiers with invalid values. I am not sure if that is a problem. Otherwise codegen would be corrupted (it was before for empty values already. That's also fixed with that).

I also had to change modifiers to be an Expression instead of SimpleExpression so that I can use parseExpression with it. That means I had to cast to SimpleExpression at a few places

Fuzzyma committed:
feat(compiler-core): allow directive modifiers to be dynamic (fix #8281)

edison1105 commented:
@Fuzzyma 
fixed by merging main into the minor branch.

edison1105 commented:
> I developed on top of the main branch but the guidelines say to submit against minor. Now I have this extra commits in my history. Not sure how to handle this

fixed by merging main into the minor branch.

edison1105 commented:
Thank you very much for your contribution. This PR is really great. I did a quick test and found the following issues.

- ### v-model dynamic modifiers on component
the following template
```html
<template>
  <Comp v-model.[mod]="msg"/>
</template>
```
should be compiled to:
```diff
_createVNode($setup["Comp"], {
      modelValue: $setup.msg,
      "onUpdate:modelValue": _cache[1] || (_cache[1] = $event => (($setup.msg) = $event)),
-      modelModifiers: { "$setup.mod": true }
+      modelModifiers: Object.assign({}, $setup.mod)
    }, null, 8 /* PROPS */, ["modelValue"])
```
---
- ### v-model static modifiers + dynamic modifiers on component 
the following template
```html
<template>
 <Comp v-model.number.[mod]="msg"/>
</template>
```
should be compiled to:
```diff
_createVNode($setup["Comp"], {
      modelValue: $setup.msg,
      "onUpdate:modelValue": _cache[1] || (_cache[1] = $event => (($setup.msg) = $event)),
-      modelModifiers: { number: true, "$setup.mod": true }
+      modelModifiers: Object.assign({number:true}, $setup.mod)
    }, null, 8 /* PROPS */, ["modelValue"])
```
---
- ### v-model dynamic modifiers with plain object on component
the following template should not cause a compilation error. see [Playground with this PR](https://deploy-preview-12913--vue-sfc-playground.netlify.app/#eNp9UslOwzAQ/RXLl4JUkgOcSqgEqAeQWATcag4hmQQXb/ISKkX5d8au2oatJ3vmvRm/meeeXhqTdQHojBaustx44sAHM2eKS6OtJ9daGtJYLckky2MQ6ZNzpop8U4BUDDxII0oPGBFSpKLuROoaRLbsiQryDeyMeBuADK8XjErXMpoju8hHpXRKvau0anibrZxWKKuPDRmtsCMXYB+M51o5RmckIRErhdCftykXH5hu89U7VB9/5FduHXOMPlpwYDtgdIf50rbgN/Di+R7WeN+BOE8QyD4APoHTIkSNG9pVUDXKHvGS2pu0XK7aF7dYe1BuO1QUGplD4jOKy47L/G/0vdzT7CzVMTXgFrdGRV+/W1Pzbt73XSkCDEORxzDluTLB7z3DgzccrItmJTaj5JdfGP74NWie82SZKqajJqSGhiu4i92Pjke/hw5ff9vaJQ==)
```html
<Comp v-model.[{ number: true }]="msg"/>
```

Fuzzyma commented:
@edison1105 thanks for the super fast feedback. Am I correct to assume, that your failing cases are specific to the v-model directive or does it happen with any other? (it shouldnt)

For the last error: When I tried this with the template-explorer it correctly showed the error `End bracket for dynamic directive modifier was not found. Note that dynamic directive modifier cannot contain spaces.` (like the one in dynamic directive args also does). The corresponding code is in the tokenizer:

https://github.com/vuejs/core/pull/12913/files#diff-25ca5a044c54b773d27122b99730a0a06343b0b058d087c4187ee778c5e87e87R760-R773

It works for other directives. But for v-model the "value required" error shadows the error from the whitespace. Not sure what to do here. Any ideas?

I am not against allowing spaces in dynamic args and modifiers but I am pretty sure it would break formatters and syntax highlighters. I guess thats also the reason why its not supported.

I would have written a test for that as well but I couldnt find where to put it.

Fuzzyma committed:
fix: special handling for modelModifiers

Fuzzyma committed:
chore: comment

Fuzzyma commented:
I have a bit of duplicated code now. Let me know if I should refactor that and if so, where I put helper functions for such things

edison1105 commented on packages/compiler-core/src/transforms/transformElement.ts:
```diff
@@ -909,15 +913,51 @@ export function buildDirectiveArgs(
+        staticMods.push(modifier)
+      } else {
+        // Object expression must be first to avoid null or undefined
+        if (staticMods.length || i === 0) {
```
L930-L937 is the same as L946-L953. Maybe it should be extracted as a method for reusing.

edison1105 commented on packages/compiler-core/src/transforms/vModel.ts:
```diff
@@ -130,26 +134,80 @@ export const transformModel: DirectiveTransform = (dir, node, context) => {
-      ),
-    )
+
+    for (let i = 0; i < dir.modifiers.length; i++) {
```
The code here seems to be the same as in transformsElement.ts and should be considered for reuse

edison1105 commented:
> I have a bit of duplicated code now. Let me know if I should refactor that and if so, where I put helper functions for such things

We can just put them in `transformElement.ts`, which will be fine.

Fuzzyma committed:
refactor: deduplicate logic

edison1105 commented:
/ecosystem-ci run

edison1105 commented:
Thanks @Fuzzyma 
Nick work and LGTM~

Fuzzyma commented:
Seems like I broke some runtime directive in vuetify :D. Have to check that...
language-tools is mostly ast problems i think.

Guess I have some more work to do. What is the process for this? Do I open a PR in language tools to support the new ast shape and get that merged before this PR gets merged?

Fuzzyma commented:
@edison1105 can i somehow checkout this PR ins vue-sfc to test it easily?

dsonet commented on packages/compiler-core/src/parser.ts:
```diff
@@ -274,7 +274,50 @@ const tokenizer = new Tokenizer(stack, {
+      )
+
+      if (!exp.ast && !exp.content.trim()) {
+        if (isStatic) {
+          emitError(
+            ErrorCodes.X_MISSING_DIRECTIVE_MODIFIER_NAME,
+            exp.loc.start.offset,
+          )
+        } else {
+          emitError(
+            ErrorCodes.X_MISSING_DYNAMIC_DIRECTIVE_MODIFIER_VALUE,
+            exp.loc.start.offset,
+          )
+        }
```
```suggestion
        emitError(
          isStatic ? ErrorCodes.X_MISSING_DIRECTIVE_MODIFIER_NAME : 
            ErrorCodes.X_MISSING_DYNAMIC_DIRECTIVE_MODIFIER_VALUE,
          exp.loc.start.offset,
        )
```

Fuzzyma committed:
refactor: parser.ts

Co-authored-by: Jian Zhang <dsonet@msn.com>

sadeghbarati commented:
Hi, Thanks for this PR 👍 

Can you check this Stackblitz playground to see if it's okay to pass down **pre-defined** `vModelText` **modifiers** like

- trim
- number
- lazy

 To the custom Input component?
 
 
 https://stackblitz.com/edit/vitejs-vite-cga23nxl

Fuzzyma commented:
@sadeghbarati what exactly should we test in that example? It seems to work just fine, right? I need more context

Fuzzyma commented:
well it does work so i dont see anything bad with doing that. However, this is not the right place to discuss this!

skirtles-code commented:
This feedback might be better placed on an RFC. I'm not aware of an RFC for this feature, so I've added it here instead.

---

If I've understood correctly, this PR only adds compiler support, but it doesn't account for the impact at runtime. Perhaps that was the intention, but I don't see this discussed anywhere. I think one of the main reasons why this feature hasn't been added earlier is because making the modifiers dynamic poses various challenges at runtime.

For example, `defineModel` doesn't support changing the modifiers:

- [Playground](https://deploy-preview-12913--vue-sfc-playground.netlify.app/#eNqFUk1PAjEQ/Su1FzDB3YOeCJAoIVET1KiJB8sB2Vkodtum7SJks//dafeDNSF42bR9b9682XkFvdU62uVAh3RkV4ZrRyy4XE+Y5JlWxpGCGEhJSVKjMtJDaq+F5ofphoukhqK4vns9JDG5UtI6ktk1GXuRfu8ehFDkQxmRXPQuW4JKbM0oiDM8G+I3B1IiYxRXrtAPXhxkWiwd4I2QUdN+d4UKIKJPL7QYM4odGSVxYBVFMFCWoYRLnTviDhqQttrA6vtL7ZFbS/ha1Ii8CUYnodib89WjuNOdDqiz6D7l62hrlcTfV/gGKKoyzQWYZ+04TsfokATEY0uc/ucxvPkBB817MHLifWvR2xAPLwYsmB0w2mJuadbgKnj29gR7PLcgms4Fss+Ar2CVyL3HinaXywRtd3jB7UPYNJfrdzvbO5C2Gcob9cwy8BnFnU/PjH60ex3dhDomS/yLnciciGCVD79XEAO/Cp5yMHaBaUkg5RLmHun/k5OE78Kh3X83MLVitXkQx61XQBOcuBL5G4LyF53CGbQ=)

```js
const [model, modifiers] = defineModel()
```

The value of `modifiers` is captured when the component is created and won't update if new modifiers are passed from the parent. In the example above, checking the checkbox won't change anything in the child.

But rewriting it to avoid `defineModel` only partially fixes the problem:

- [Playground](https://deploy-preview-12913--vue-sfc-playground.netlify.app/#eNqFU01v1DAQ/SvGl6TSkhzgFG0rQbUSIJVWFMGh6WGbTFIvjm35I2wV5b8zdtZJQNVyiTIzb968Gc8M9INSWe+AFnRrKs2UJQasU1elYJ2S2pKBaGjISBotO5IgNJlDNy/Xz4zXp1CWn2zPh6BSVFIYSzrTkktPkiafgHNJfkrN6zfJxQyQtTkhBmI16wr8OiAjIrb5pAr1oGGhU3xvAS1CtrF8/xYZgGcPnujxsqRYsaQkD6hhCALGMaQwoZwl9kUBwqpnqH49ySNiTxQ+FzkyL6KkVyHZi/PZ23xVnW6oNai+YW12MFLg+AZfAEllpxgHfassw+5KWpAQ8bE9dv/7S/D5BjfRH4S84j8Y1Fbgz50GA7qHks4xu9ct2Cm8u/8KR/yfgyjacUSfCX4DI7nzGifYRydqlL3CBbWfw0sz0X43u6MFYWJTXqhHjgFfUnzz6zOtL3LfZe9DXilGnOJqZc6toJ+qs1D/s4dxg5SWyq9QDQ0TcOetNJQOr/pjzx0U5B5fVbRBbnDfyJo1DLQpyO3TASqLinDjIid0zM6UOzRM+pA4VeMCFAtt8rhKCW7MiWonDdh3ehEnobEzLSbB2UKzTBJbT3vvmlO8kPSVyhsy4eI0/3MtNevDz3wF67NZTSMckPcsB7CKxkPKJ7q/j2L8A8Jeel0=)

The modifiers now make it to the child component, but passing them to the `<input>` won't help because `trim` is applied in the `created` hook of `vModelText`.

This is one specific example, but the problem is a general one: all directives will now need to account for modifiers changing, as well as any components that use `v-model` modifiers. While this may have been theoretically possible previously via render functions, I don't think there was any real expectation that modifiers could change beyond their initial values. I think introducing template syntax for dynamic modifiers will change that expectation.

---

There are also some modifiers that are resolved at compile-time, which can't work if they're dynamic. e.g.:

- [Playground](https://deploy-preview-12913--vue-sfc-playground.netlify.app/#eNp9UctOwzAQ/BXLl4JUpQc4VSnioR7gAAi4YQ5Rsg1uHdvyrtuiKP/OOqGlqqrevDOzszNyK++8z9YR5FTmWAbtSSBQ9DfKls4iicZVKGaiFcVUUIggOmXzySBlEQ8EjTcFAU9C5Nr6SOJ2BT+V29ismCmZjJyBzLj6YrTQAWl0qeRJ+Wc693W8g8Bj9beUTw4OyrEkZHKh62yJznKNNvkmg8ZrA+HFk2YvJaeiZxJXGOM2Tz2WKo13ePkN5eoEvsRtwpR8DYAQ1qDknqMi1EADPX9/hi2/9yS3iYbVZ8g34J4xZRxk99FWHPtA16d9bLwLpG39gfMtgcVdqRQ0KbteryT/5cOZ6v9xr7Lrfk/ZTna/TPytJA==)

```html
<input @keydown.[mods]="console.log('second')">
```

Here `mods` might be a key modifier, or one of the general modifiers like `stop` or `prevent`. To get this working, the logic that's currently handled in the compiler would need to be implemented at runtime.

Fuzzyma commented:
@skirtles-code thank you for the in-depth feedback! You are right. I totally forgot about the RFC procedure and jumped the gun.
If you want, I can take a step back and prepare an RFC for this feature to let it go the normal route.

Meanwhile:
I do agree with you, that modifiers were always perceived as static and that this PR might "damage" that perception.
I think that modifiers should stay static for the issues you mentioned and it should be made very clear in the docs, that this feature is intended only for modifiers to be dynamic and not reactive.

_However, just in case, reactive modifiers should become a thing, I could see defineModel work exactly like defineProps in that it returns reactive variables that are tracked by the compiler to stay in a reactive context. All the other problems aside, this could be a fix for this specific problem (again, i dont think its a good idea)_

> There are also some modifiers that are resolved at compile-time, which can't work if they're dynamic.

Yes, that is a problem indeed (Beside the fact, that this is still not correctly compiled as of right now).
The only solution I can think of my head is to treat all dynamic modifiers as both - keys and guards. Since no key is named "prevent" or "stop" and no guard is named "a", this should work because unknown keys and guards are dropped.

Basically let it compile to this:

```ts
_withKeys(_withModifiers($event => (console.log('second')), [mods]), [mods]))
```

What do you think? It is runtime overhead but specifically only for dynamic modifiers.

Fuzzyma committed:
feat(compiler): support dynamic directive modifiers for v-model

- v-on is not supported for now
- vapor mode support was added
- better type guards

Fuzzyma commented:
I updated the PR to limit the change to v-model only and skip v-on support for now. Namely because of the discussed problems and the need for runtime code in a lot of cases.

Now, it only is up for discussion if dynamic modifiers should be reactive. My clear answer is no. Because thats also not what happens when using dynamic modifiers in a render function and I would argue it breaks user expectations
